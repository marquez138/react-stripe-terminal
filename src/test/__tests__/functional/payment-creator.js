import React from 'react';
import rp from 'request-promise';
import { TestSetup, TestComponent } from '../../helpers/test-setup';
import { StripeTerminalProvider, injectStripeTerminal } from '../../../index';
import TestRenderer from 'react-test-renderer';

const createPaymentIntent = async ({ amount, currency = 'usd' }) => {
    const response = await rp.post(
        `https://api.stripe.com/v1/payment_intents?amount=${amount}&currency=${currency}&allowed_source_types[]=card_present&capture_method=manual`,
        {
            headers: {
                Authorization: `Bearer ${TestSetup.TEST_API_KEY}`,
            },
        }
    );

    const result = JSON.parse(response);
    return result.client_secret;
};

class PaymentComponent extends TestComponent {
    componentDidUpdate() {
        this.props.onVerifyState(this.props.stripeTerminal);
    }
    render() {
        const { stripeTerminal } = this.props;
        return <div>Payment Status: {stripeTerminal.payment.status}</div>;
    }
}
describe('payment-creator', () => {
    /**
     * Creating payments takes a little longer so test timeouts here are 10s
     */
    describe('createPayment', () => {
        it('creates a payment', async () => {
            const TestApp = injectStripeTerminal(PaymentComponent);
            let stripeTerminal, piHandlerCalled;
            let renderer, lastPaymentState;
            const testElement = (
                <StripeTerminalProvider
                    onFetchConnectionToken={TestSetup.FetchConnectionToken}
                    onPaymentIntentRequest={async options => {
                        piHandlerCalled = true;
                        const pi = await createPaymentIntent({ ...options });
                        return pi;
                    }}
                    currency={'USD'}
                    taxRate={0.07}
                >
                    <TestApp
                        onVerifyState={stripeTerminal => {
                            if (!renderer) {
                                return;
                            }
                            if (
                                stripeTerminal.payment.status !==
                                lastPaymentState
                            ) {
                                lastPaymentState =
                                    stripeTerminal.payment.status;
                                expect(renderer.toJSON()).toMatchSnapshot();
                            }
                        }}
                        stripeTerminalTestHook={t => {
                            stripeTerminal = t;
                        }}
                    />
                </StripeTerminalProvider>
            );
            renderer = TestRenderer.create(testElement);
            expect(renderer.toJSON()).toMatchSnapshot();

            const readers = await stripeTerminal.discoverReaders({
                method: 'simulated',
            });
            let conn = await stripeTerminal.connectReader(readers[0]);

            let pi = await stripeTerminal.createPayment({
                paymentIntentOptions: {
                    amount: 100,
                    currency: 'usd',
                },
            });
            renderer.update(testElement);
            expect(piHandlerCalled).toBe(true);
            expect(typeof pi).toBe('object');
            await stripeTerminal.disconnectReader();
        }, 10000);
    });

    describe('cancelCreatePayment', () => {
        it('cancels an in-flight payment', async () => {
            const TestApp = injectStripeTerminal(PaymentComponent);
            let stripeTerminal, piHandlerCalled;
            let renderer, lastPaymentState, cancelled;
            const testElement = (
                <StripeTerminalProvider
                    onFetchConnectionToken={TestSetup.FetchConnectionToken}
                    onPaymentIntentRequest={async options => {
                        piHandlerCalled = true;
                        const pi = await createPaymentIntent({ ...options });
                        return pi;
                    }}
                    currency={'USD'}
                    taxRate={0.07}
                >
                    <TestApp
                        onVerifyState={async stripeTerminal => {
                            if (!renderer) {
                                return;
                            }
                            if (
                                lastPaymentState !==
                                stripeTerminal.payment.status
                            ) {
                                expect(renderer.toJSON()).toMatchSnapshot();
                                lastPaymentState =
                                    stripeTerminal.payment.status;
                            }

                            if (
                                stripeTerminal.payment.status ===
                                    'collecting_payment_method' &&
                                !cancelled
                            ) {
                                cancelled = true;
                                try {
                                    await stripeTerminal.cancelCreatePayment();
                                } catch (e) {
                                    console.error(e);
                                }
                            }
                        }}
                        stripeTerminalTestHook={t => {
                            stripeTerminal = t;
                        }}
                    />
                </StripeTerminalProvider>
            );
            renderer = TestRenderer.create(testElement);
            expect(renderer.toJSON()).toMatchSnapshot();

            const readers = await stripeTerminal.discoverReaders({
                method: 'simulated',
            });
            await stripeTerminal.connectReader(readers[0]);

            await stripeTerminal.createPayment({
                paymentIntentOptions: {
                    amount: 100,
                    currency: 'usd',
                },
            });

            expect(piHandlerCalled).toBe(true);
            renderer.update(testElement);
            expect(renderer.toJSON()).toMatchSnapshot();
            await stripeTerminal.disconnectReader();
        }, 10000);
    });
});
