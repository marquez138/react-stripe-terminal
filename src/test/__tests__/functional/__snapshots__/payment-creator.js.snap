// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payment-creator cancelCreatePayment cancels an in-flight payment 1`] = `
<div>
  Payment Status: 
  ready
</div>
`;

exports[`payment-creator cancelCreatePayment cancels an in-flight payment 2`] = `
<div>
  Payment Status: 
  ready
</div>
`;

exports[`payment-creator cancelCreatePayment cancels an in-flight payment 3`] = `
<div>
  Payment Status: 
  collecting_payment_method
</div>
`;

exports[`payment-creator cancelCreatePayment cancels an in-flight payment 4`] = `
<div>
  Payment Status: 
  ready
</div>
`;

exports[`payment-creator cancelCreatePayment cancels an in-flight payment 5`] = `
<div>
  Payment Status: 
  ready
</div>
`;

exports[`payment-creator createPayment creates a payment 1`] = `
<div>
  Payment Status: 
  ready
</div>
`;

exports[`payment-creator createPayment creates a payment 2`] = `
<div>
  Payment Status: 
  ready
</div>
`;

exports[`payment-creator createPayment creates a payment 3`] = `
<div>
  Payment Status: 
  collecting_payment_method
</div>
`;

exports[`payment-creator createPayment creates a payment 4`] = `
<div>
  Payment Status: 
  ready
</div>
`;

exports[`payment-creator createPayment creates a payment 5`] = `
<div>
  Payment Status: 
  confirming_payment_intent
</div>
`;

exports[`payment-creator createPayment creates a payment 6`] = `
<div>
  Payment Status: 
  ready
</div>
`;
